sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,835 @@
+ */
+
+/*
+ * virt IO FE driver
+ *
+ * The SOF driver thinks this driver is another audio DSP, however the calls
+ * made by the SOF driver core do not directly go to HW, but over a virtIO
+ * message Q to the virtIO BE driver.
+ *
+ * The virtIO message Q uses the *exact* same IPC structures as we currently
+ * use in the mailbox.
+ *
+ * The mailbox IO and TX/RX msg functions below do IO on the virt IO Q.
+ */
+
+#include <linux/completion.h>
+#include <linux/device.h>
+#include <linux/firmware.h>
+#include <linux/interrupt.h>
+#include <linux/module.h>
+#include <linux/scatterlist.h>
+#include <linux/virtio.h>
+#include <linux/virtio_config.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_ring.h>
+#include <sound/sof.h>
+#include <sound/sof/virtio.h>
+
+#include "ops.h"
+#include "sof-audio.h"
+#include "sof-priv.h"
+
+#define SOF_COMPONENT_NAME	"sof-audio-component"
+/* 600ms for VirtQ IPC */
+#define DATA_TIMEOUT_MS		1000
+
+static const char *const sof_vq_names[SOF_VIRTIO_NUM_OF_VQS] = {
+	SOF_VIRTIO_IPC_CMD_VQ_NAME,
+	SOF_VIRTIO_IPC_PSN_VQ_NAME,
+	SOF_VIRTIO_DATA_VQ_NAME,
+};
+
+struct sof_vfe {
+	struct snd_sof_dev *sdev;
+
+	/* IPC cmd from frontend to backend */
+	struct virtqueue *ipc_cmd_vq;
+
+	/* IPC position update from backend to frontend */
+	struct virtqueue *ipc_psn_vq;
+
+	/* audio data in both directions */
+	struct virtqueue *data_vq;
+
+	/* position update work */
+	struct work_struct posn_update_work;
+
+	/* current pending cmd message */
+	struct snd_sof_ipc_msg *msg;
+
+	/* current and pending notification */
+	struct snd_sof_ipc_msg *not;
+	struct sof_ipc_stream_posn posn;
+
+	/*
+	 * IPC messages are blocked. "true" if the DSP hasn't been reset and
+	 * therefore we don't have to re-send our topology.
+	 */
+	bool block_ipc;
+	struct sof_vfe_ipc_tplg_resp tplg;
+
+	struct completion completion;
+	spinlock_t vq_lock;
+
+	/* A shared capture / playback virtual queue data buffer */
+	union {
+		struct dsp_sof_data_req data_req;
+		struct dsp_sof_data_resp data_resp;
+	};
+
+	/* Headers, used as a playback response or capture request */
+	union {
+		u8 hdr_req[HDR_SIZE_REQ];
+		u8 hdr_resp[HDR_SIZE_RESP];
+	};
+};
+
+/*
+ * IPC Firmware ready.
+ */
+static int sof_vfe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
+{
+	return 0;
+};
+
+/* used to send IPC to BE */
+static int sof_vfe_send_msg(struct snd_sof_dev *sdev,
+			    struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
+	size_t msg_size = msg->msg_size;
+	void *msg_data = msg->msg_data;
+	int ret;
+
+	if (vfe->block_ipc) {
+		struct sof_ipc_reply *reply = msg->reply_data;
+
+		reply->error = 0;
+		msg->reply_error = reply->error;
+		/*
+		 * No need to take .ipc_lock: we return > 0, so
+		 * sof_ipc_tx_message_unlocked() won't overwrite .ipc_complete
+		 */
+		msg->ipc_complete = true;
+		wake_up(&msg->waitq);
+
+		return 1;
```

Comment:
@lyakh do you mind explaining whats happening here? We're trying to send an IPC to vBE?

Reply:
@ranj063 this is the `.block_ipc = true` case, it means, that the guest is waking up after a runtime suspend, but the host wasn't suspended during that time. So, the guest tries to send its topology to the DSP again, which isn't needed in this case. So this code just pretends that the IPC was sent successfully. The opposite case is when both the guest and the host were runtime suspended, so the DSP was off and it needs the topology to be resent.

Reply:
@lyakh cool thanks! this needs to be somewhere in the comments so other people wont ask the same questions

Reply:
@ranj063 it is already there:
```
	/*
	 * IPC messages are blocked. "true" if the DSP hasn't been reset and
	 * therefore we don't have to re-send our topology.
	 */
	bool block_ipc;
```